### DISCUSSION TITLE
How reproduce few iperf3 errors related to test initialization and stream id

### DISCUSSION THREAD

**Discussion by Bala-101423** [esnet/iperf#1813] [General]
Hi, Can you please tell how to re produce below iperf3 based errors on linux/windows machine.
IENEWTEST = 100,        // Unable to create a new test (check perror)
IEINITTEST = 101,       // Test initialization failed (check perror)
 /* Stream errors */
    IECREATESTREAM = 200,   // Unable to create a new stream (check herror/perror)
    ...
    IESTREAMID = 208,       // Stream has invalid ID
    /* Timer errors */
    IENEWTIMER = 300,       // Unable to create new timer (check perror)
    IEUPDATETIMER = 301,    // Unable to update timer (check perror)

**Comment by davidBar-On** [comment]
I am not sure it is practical to try causing all errors.  There are set in many places and the number of tests to produce them may be huge.

In general, for some of the errors you can kill the server or client in different phases of the test.  For others, you can disconnect the link between the server and client in different test phases (by pulling out network cable, turn off WiFi, etc.).

For other errors (probably the most of them) you have to change the code to cause the error.  One type of change that will cause some (few?) is to change the contents of written messages or read messages (or don't send messages to cause timeout).

For other errors you need find where in the code these errors are issued and change the code at that place so it will issue the error.  E.g. `IESETNODELAY` is set in 4 places.  So you will have to make a change and run a test 4 times.  Exampel for the change, in `iperf_client_api.c` change:
```c
    if (setsockopt(test->ctrl_sck, IPPROTO_TCP, TCP_NODELAY, (char *) &flag, sizeof(int))) {
        i_errno = IESETNODELAY;
        return -1;
    }
```

to (note the added `!`):

```c
    if (!setsockopt(test->ctrl_sck, IPPROTO_TCP, TCP_NODELAY, (char *) &flag, sizeof(int))) {
        i_errno = IESETNODELAY;
        return -1;
    }
```

**Reply by Bala-101423** [reply]
Hi David, Thank you for the response and inputs.
